app/gui2/src/components/GraphEditor/GraphVisualization.vue

```diff
@@ -117,10 +123,45 @@ watchEffect(async () => {
   }
 })
 
+const isBelowToolbar = ref(false)
+let width = ref<number | null>(null)
+let height = ref(150)
+
+watchEffect(() =>
+  emit(
+    'update:rect',
+    new Rect(
+      props.nodePosition,
+      new Vec2(
+        width.value ?? props.nodeSize.x,
+        height.value + (isBelowToolbar.value ? TOP_WITH_TOOLBAR_PX : TOP_WITHOUT_TOOLBAR_PX),
+      ),
+    ),
+  ),
+)
+
+onUnmounted(() => emit('update:rect', undefined))
+
 provideVisualizationConfig({
   fullscreen: false,
-  width: null,
-  height: 150,
+  get width() {
+    return width.value
+  },
+  set width(value) {
+    width.value = value
+  },
+  get height() {
+    return height.value
+  },
+  set height(value) {
+    height.value = value
+  },
+  get isBelowToolbar() {
+    return isBelowToolbar.value
+  },
+  set isBelowToolbar(value) {
+    isBelowToolbar.value = value
+  },
```

Comment:
I don't like the idea that config became a two-directional way of exchanging information. Most fields are just passed from `GraphVisualization` to concrete instances, except `isBelowToolbar` which is instead settled by the instances, and here `GraphVisualization` read it. This should be called "model" not config. Or we should split it into two objects (config and model). Or propagate isBelowToolbar through event?

BTW why we use providers here instead of v-models? Why `hide` and `updateType` are passed through so-called "config", but "updatePreprocessor" is a normal event?

cc @Frizi maybe you have some opinion about the above?